internal/stream/stream_test.go

```diff
@@ -250,3 +251,109 @@ func TestPushStream_Stop(t *testing.T) {
+func TestNewPushStream_Failure(t *testing.T) {
+	ctx := context.Background()
+	ctrl := gomock.NewController(t)
+	subscriptionCoreMock := mocks1.NewMockICore(ctrl)
+	subscriberCoreMock := mocks2.NewMockICore(ctrl)
+	workerID := uuid.New().String()
+	httpConfig := &httpclient.Config{}
+	subscriptionCoreMock.EXPECT().Get(gomock.Any(), subName).Return(nil, errors.New("Test Error"))
+	got, err := NewPushStream(ctx, workerID, subName, subscriptionCoreMock, subscriberCoreMock, httpConfig)
+	assert.Nil(t, got)
+	assert.NotNil(t, err)
+}
+
+func TestStart_RestartSubscriber(t *testing.T) {
+	ctx := context.Background()
+	ctrl := gomock.NewController(t)
+
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte(`{"IsSuccess": true}`))
+	}))
+
+	defer s.Close()
+
+	tests := []struct {
+		name      string
+		endpoint  string
+		isSuccess bool
+	}{
+		{
+			name:      "Run Push Stream Restart Subscriber",
+			endpoint:  s.URL,
+			isSuccess: true,
+		},
+	}
+
+	for _, test := range tests {
+		ps, sub := getMockPushStreamRestart(ctx, ctrl, test.endpoint)
+		go ps.Start()
+		<-time.NewTicker(10 * time.Millisecond).C
+		if ps.subs != nil {
+			ps.subs.GetErrorChannel() <- errors.New("Subscriber Test Error")
+		}
+		select {
+		case <-time.NewTicker(1 * time.Second).C:
+			assert.NotEqual(t, ps.subs, sub)
+			ps.Stop()
+		}
+	}
+}
+
+func getMockPushStreamRestart(ctx context.Context, ctrl *gomock.Controller, endpoint string) (*PushStream, *mocks2.MockICore) {
+	subscriptionCoreMock := mocks1.NewMockICore(ctrl)
+	subscriberCoreMock := mocks2.NewMockICore(ctrl)
+	subModel := getMockSubModel(endpoint)
+	subscriptionCoreMock.EXPECT().Get(gomock.Any(), subName).Return(subModel, nil)
+	workerID := uuid.New().String()
+	httpConfig := &httpclient.Config{}
+	pushStream, _ := NewPushStream(ctx, workerID, subName, subscriptionCoreMock, subscriberCoreMock, httpConfig)
+	pushStream.subs = getMockSubscriberRestart(ctx, ctrl)
+	subscriberCoreMock.EXPECT().NewSubscriber(
+		ctx,
+		workerID,
+		subModel,
+		defaultTimeoutMs,
+		defaultMaxOutstandingMsgs,
+		defaultMaxOuttandingBytes,
+		gomock.AssignableToTypeOf(make(chan *subscriber.PullRequest)),
+		gomock.AssignableToTypeOf(make(chan *subscriber.AckMessage)),
+		gomock.AssignableToTypeOf(make(chan *subscriber.ModAckMessage))).Return(getMockSubscriberRestart(ctx, ctrl), nil).Times(2)
+	return pushStream, subscriberCoreMock
+}
+
+func getMockSubscriberRestart(ctx context.Context, ctrl *gomock.Controller) *mocks3.MockISubscriber {
+	subscriberMock := mocks3.NewMockISubscriber(ctrl)
+	reqCh := make(chan *subscriber.PullRequest, 1)
+	resCh := make(chan *metrov1.PullResponse)
+	cancelChan := make(chan bool)
+
+	go func() {
+		counter := 0
+		for {
+			select {
+			case <-reqCh:
+				if counter < 1 {
+					messages := getMockResponseMessages()
+					resCh <- &metrov1.PullResponse{
+						ReceivedMessages: messages,
+					}
+					counter++
+				}
+				resCh <- &metrov1.PullResponse{}
```

Comment:
this is a blocking.. resChan is unbuffered channel, and it will block the execution if stream is not listening on this channel. Please wrap this one inside else block.

Reply:
added in else.